gax/backoff_policy.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_API_CODEGEN_GAPIC_BACKOFF_POLICY_H_
+#define GOOGLE_API_CODEGEN_GAPIC_BACKOFF_POLICY_H_
+
+#include <algorithm>
+#include <chrono>
+#include <memory>
+#include <random>
+
+namespace google {
+namespace gapic {
+namespace gax {
+
+class BackoffPolicy {
+ public:
+  virtual ~BackoffPolicy() = default;
+
+  /**
+   * Handle an operation completion.
+   *
+   * This function is called when an operation has failed and needs to be retried.
+   * The decision to retry or not is handled by other policies.
+   *
+   * @return the delay to wait before the next retry attempt.
+   */
+  virtual std::chrono::microseconds OnCompletion() = 0;
+
+  /**
+   * Return a new copy of this object.
+   */
+  virtual std::unique_ptr<BackoffPolicy> clone() const = 0;
+
+  // Used in testing
+  virtual std::chrono::microseconds CurrentDelay() const = 0;
+  virtual std::chrono::microseconds MaximumDelay() const = 0;
+};
+
+/**
+ * Implements a truncated exponential backoff with randomization policy.
+ *
+ * This policy implements the truncated exponential backoff policy for
+ * retrying operations. After a request fails, and subject to a separate
+ * retry policy, the client library will wait for an initial delay before
+ * trying again. If the second attempt fails the delay time is increased by a factor
+ * of 2. The delay time growth stops at a maximum delay wait time.
+ * The policy also randomizes the delay each time, to avoid
+ * [thundering herd
+ * problem](https://en.wikipedia.org/wiki/Thundering_herd_problem).
+ *
+ * Note: The random number generator used when calculating backoff time in
+ *       OnCompletion is lazily created. Furthermore, GeneratorFactory::Generator
+ *       may touch mutable shared state.
+ *       As a result, OnCompletion is NOT thread-safe.
+ *       All other methods, including clone, are thread-safe.
+ *       ExponentialBackoffPolicy is therefore thread-compatible.
+ */
+template <typename GeneratorFactory>
+class ExponentialBackoffPolicy : BackoffPolicy {
```

Comment:
OnCompletion references the GeneratorFactory template parameter and therefor has to live in a header. This could be obfuscated more than it currently is, but it's still something that other compilation units would have to see somehow.
The constructors could certainly be moved to a .cc file, but I'm not sure it's worth it.

Reply:
Resolved offline to just make an rng in place without templates.